### PR TITLE
Informative git status on prompt

### PR DESCRIPTION
**Pull request by blahgeek**
screen shot 2016-09-03 at 01 14 57

Much more informative git status!

Should this go into xonsh core or remains xontrib? And, can I add documentations in `docs/` (as this is a xontrib...) ?

Thanks for reviewing!

### CODE REVIEW

**Comment by melund** [comment]
Nice. I want it in core if it isn't slower :)

File: xontrib/gitstatus.xsh [hunk 1]
```diff
@@ -0,0 +1,140 @@
+'''informative git status for prompt'''
```
Comment (scopatz) [inline xontrib/gitstatus.xsh]: PEP78 triple double quotes

File: xontrib/gitstatus.xsh [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import builtins
+import functools
+import subprocess
+
+__all__ = []
+
+_env = builtins.__xonsh_env__
+
+check_output = functools.partial(subprocess.check_output,
+                                 env=_env.detype(),
```
Comment (scopatz) [inline xontrib/gitstatus.xsh]: I don't think this is valid at the global scope.  The env needs to be detyped each time, right before the call

File: xontrib/gitstatus.xsh [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                                 stderr=subprocess.DEVNULL,
+                                 timeout=_env['VC_BRANCH_TIMEOUT'],
```
Comment (scopatz) [inline xontrib/gitstatus.xsh]: Same here. What if a user changes the timeout?

**Comment by scopatz** [comment]
I agree. This would be great to have in the core!  Thanks for putting this in @blahgeek!

**Comment by blahgeek** [comment]
It's now in core.

I alse moved all `FORMATTER_DICT` related functions to `xonsh/prompt/` sub-package, created a new amalgamate.

**Comment by melund** [comment]
Thanks for the nice work @blahgeek. I agree that a restructure was probably needed here. The new structure also seems more logical.  But it  is with mixed feelings. I really hate that I can't no longer remember where things are located or easily see who coded what. The former can probably be fixed by reviewing this but I don't have time to look at this in details now.Will try to find time later this weekend

File: xonsh/prompt/base.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
```
Comment (laerus) [inline xonsh/prompt/base.py]: No hashbang here, this will never be executed as a standalone script right?

File: xonsh/prompt/cwd.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# Created by i@BlahGeek.com at 2016-09-03
+
```
Comment (laerus) [inline xonsh/prompt/cwd.py]: Module docstrings would be nice aswell

File: xonsh/prompt/gitstatus.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+def check_output(*args, **kwargs):
+    kwargs.update(dict(env=builtins.__xonsh_env__.detype(),
+                       stderr=subprocess.DEVNULL,
+                       timeout=builtins.__xonsh_env__['VC_BRANCH_TIMEOUT'],
+                       universal_newlines=True
+                       ))
+    return subprocess.check_output(*args, **kwargs)
+
+_DEFS = {
```
Comment (laerus) [inline xonsh/prompt/gitstatus.py]: this can be a function decorated with `lazyobject` from [lazyasd](https://github.com/xonsh/xonsh/blob/master/xonsh/lazyasd.py#L126) that returns the dict like:
```
@lazyobject
def _DEFS():
    return {...}
```

File: xonsh/prompt/cwd.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# Created by i@BlahGeek.com at 2016-09-03
```
Comment (scopatz) [inline xonsh/prompt/cwd.py]: We don't have a custom of signing files in xonsh. If you could remove this line, that would be most appreciated!

File: xonsh/prompt/base.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+# -*- coding: utf-8 -*-
+
+import os
+import socket
+
+from xonsh.lazyasd import LazyObject
+from xonsh.tools import is_superuser
+from xonsh.platform import ON_WINDOWS
```
Comment (scopatz) [inline xonsh/prompt/base.py]: You should not do `from ... import` for modules outside of the subpackage.  This kinds of imports cannot be performed lazily.

File: xonsh/prompt/cwd.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+import os
+import shutil
+import builtins
+
+from xonsh.tools import get_sep
+from xonsh.platform import ON_WINDOWS
```
Comment (scopatz) [inline xonsh/prompt/cwd.py]: Same here

**Comment by scopatz** [comment]
Thanks for this awesome change @blahgeek!   In addition to the inline notes, the autodoc hooks in the `docs/api` folder now need to be updated too.

**Comment by blahgeek** [comment]
Sure. I just removed those inline comments, just old habits and editor plugins, never mind :P 

I would take a look at `docs/api` and the imports issues later.

**Comment by blahgeek** [comment]
`from ... import ...` statements are removed. `docs/api/prompt` added. Any suggestions?

File: docs/api/prompt/index.rst [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+=================
+Prompt API
+=================
+Functions in ``FORMATTER_DICT`` for prompt.
+
+.. toctree::
+    :maxdepth: 1
+
+    gitstatus
+    vc_branch
```
Comment (scopatz) [inline docs/api/prompt/index.rst]: base, cwd, and env need entries and doc files too please.

File: news/gitstatus.rst [hunk 11]
```diff
@@ -0,0 +1,13 @@
+**Added:**
+
+* New ``FORMATTER_DICT`` entry ``gitstatus`` to provides informative git status
+
+**Changed:** None
+
```
Comment (scopatz) [inline news/gitstatus.rst]: Please also add a changed entry for the new prompt subpackage.

File: xonsh/environ.py [hunk 12]
```diff
@@ -45,6 +40,8 @@
     seq_to_upper_pathsep,
 )
 
+from xonsh.prompt.base import FORMATTER_DICT
```
Comment (scopatz) [inline xonsh/environ.py]: You should not do `from ... import ...` for elements outside of the current package as this will not be lazy.

File: xonsh/tracer.py [hunk 13]
```diff
@@ -12,7 +12,7 @@
 from xonsh.platform import HAS_PYGMENTS
 from xonsh.tools import DefaultNotGiven, print_color, normabspath, to_bool
 from xonsh.inspectors import find_file, getouterframes
-from xonsh.environ import _replace_home
+from xonsh.prompt.cwd import _replace_home
```
Comment (scopatz) [inline xonsh/tracer.py]: Same here, should not do `from ... import ...` for objects outside of the current subpackage.

File: xonsh/prompt/gitstatus.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    stashed = _get_stash(gitdir)
+    operations = _gitoperation(gitdir)
+
+    return (branch, num_ahead, num_behind,
+            untracked, changed, conflicts, staged, stashed,
+            operations)
+
+
+def gitstatus_prompt():
+    """Return str `[BRANCH|OPERATOR|numbers]`"""
```
Comment (scopatz) [inline xonsh/prompt/gitstatus.py]: I don't think that the return should contain the bounding `[` and `]`, because this is something that the user may wish to have / not have in their own prompt.  I don't count this as blocking and I am happy to put in a separate PR to this effect.

**Comment by scopatz** [comment]
Just a couple of more small issues, @blahgeek and then I think this is ready to go! Thanks again!

File: docs/api/prompt/index.rst (hunk 10 above)
Comment (blahgeek) [inline docs/api/prompt/index.rst]: I thought about that... but those files do not include any functions that does not starts with `_` and have docstrings, should I include them as well?

Comment (scopatz) [inline docs/api/prompt/index.rst]: No, but the should have module level docstrings and they may have public members in the future.

**Comment by scopatz** [comment]
This looks good to me! Let us know when you think it is ready!

**Comment by blahgeek** [comment]
@scopatz I think it's good to go. Thanks a million!

**Comment by scopatz** [comment]
Thanks again @blahgeek! This was a much needed refactor!

**Comment by laerus** [comment]
cool! just a question @blahgeek, can we drop `vc_branch.get_git_branch` in favor of `gitstatus.gitstatus` which seems a lot faster, or am i missing something?

**Comment by blahgeek** [comment]
I think, basically, yes. But it seems that `vc_branch.get_git_branch` handles more edge cases (on windows, when git executable do not exist, etc.) and surely we would not want to break any compatibility for existing users using `vc_branch`, so I just left the whole `vc_branch` intact for now. 

I think we can surely make more enhancement to `gitstatus` and maybe merge `gitstatus` and `vc_branch` some day in the future.
